### ISSUE TITLE
Example: "To Import a System from the SAM GUI" seems broken....

### ISSUE THREAD

**Issue by frohro** [NREL/pysam#20]
This [example](https://nrel-pysam.readthedocs.io/en/latest/Import.html) has this error:
Traceback (most recent call last):

  File "/home/frohro/SAM/2020.1.17/LEAC1_test.py", line 25, in <module>
    so.execute()

error: singleowner execution error. precheck input: variable 'ppa_price_input' required but not assigned

I'm using the beta 2020.1.17 when I do this, and PySAM 2.02.  In SAM I simply added a new Generic System, and went to Generate, and generated the .json for it.  It was untitled(1).

I'll insert my code in case it is useful.  I just copied and pasted it from the example.
```
import json
import PySAM.GenericSystem as GenericSystem
import PySAM.Singleowner as Singleowner
from PySAM.PySSC import *
ssc = PySSC()
with open("/home/frohro/SAM/2020.1.17/untitled__1_.json") as f:
    dic = json.load(f)
    gs_dat = dict_to_ssc_table(dic, "generic_system")
    so_dat = dict_to_ssc_table(dic, "singleowner")
    gs = GenericSystem.wrap(gs_dat)
    # to create GenericSystem and Singleowner combined simulation, sharing the same data
    so = Singleowner.from_existing(gs)
    so.assign(Singleowner.wrap(so_dat).export())
    gs.execute()
    so.execute()
    so.Outputs.export() # as dictionary
    print('gs.Outputs.ac_annual: ', gs.Outputs.ac_annual)
```
This issue isn't really important to me, but I thought I would report it.  The one I care about will be next.  :-)

**Comment by dguittet** [comment]
@frohro thanks for the report

What financial model did you use when creating this in SAM? This example is for the Generic System - PPA Single Owner configuration. If you created it with another financial model, `ppa_price_input` may indeed be missing.

I ran this and it worked for me (besides gs.Outputs.ac_annual being changed to annual_energy)

**Comment by frohro** [comment]
Hi Darice,
You are right.  I had the wrong model.  I see you already added the PPA part into the documentation.  Thanks for that.  I think that would have solved my issue.  To me it seems that it would be good to have an algorithm spelled out in the documentation showing a PySAM user exactly the steps he/she needs to do to duplicate the results from SAM.  Ideally this would include an example with screen shots from SAM, and links to the appropriate documentation pages, with instructions how to translate that into python, telling what each python line does, so the PySAM user can change those to be appropriate for his/her model.  From my experience so far it is kind of like this:

1. Make the model you care about in SAM
2. Got to the new model menu for the model you just created then ->Generate Code->Json for inputs
3. This will save a .json file with the name of your model, and it can be enlightening to read that file in a text editor.  You will see a dictionary of the input variable names and their values.  Sometimes it may be unclear what variable in the SAM gui corresponds to a variable in PySAM.  You can put that variable to a hopefully unique number, and search the json file for it to help you verify that you have the correct variable name in PySAM.  You also need to know those variables are for classes, and those classes can be found in the PySAM documentation for your model.  The example could show printing several of the variables from the chain of models to make clear how these variables are part of their models.  The SDKtool also gives further hints as to the variable names.
4. In python you read import all the needed modules.  I found following your example using analogy helpful.  I'm using Ubuntu Linux and haven't got your stubs working.  I think that would help.  It would be nice to have some cross platform instructions on getting those working.
5. You load the the json file
6. You determine which specific sub modules you need to run and their order by looking at [this page](https://nrel-pysam.readthedocs.io/en/latest/Configs.html)
7. You convert the dictionaries to ssc tables
8. You wrap those for each module.  I'm not sure what this is doing and why it is needed.  I'm just following the example.
9. You create system objects for each of the sub system modules you need to execute for that model.  
10. You load the wrapped data into those system objects
11. You execute each system in order (some systems need outputs from previous systems)
12. You print the variables you want to know

Can you correct me on my understanding?  If you like, I'm happy to try and put some more details in your documentation and submit it for your review.

Thanks!

Rob

**Comment by dguittet** [comment]
@frohro This is great and so helpful! The documentation is definitely very bare at the moment and thank you for working through this and writing it down. 

4) Regarding the stubs, what IDE are you using? I'm on Mac and using PyCharm, I've seen PyCharm integrate those stubs immediately after downloading. It may be a matter of seeing how your IDE supports stubs.

8) This is a workaround for the code generator not being adapted for PySAM (it's on my list!). Basically the dictionaries returned by `dict_to_ssc_table` in this [example](https://nrel-pysam.readthedocs.io/en/master/Import.html) are specific to the model type. Then those values need to be read-in by the appropriate PySAM model.

the `from_existing` function introduced in version 2.0.2 allows multiple PySAM models to use the same underlying data.

```
so = Singleowner.from_existing(gs)
so.assign(Singleowner.wrap(so_dat).export())
```
Here `so` is using the same underlying ssc table as `gs` but `gs` didn't have any of the singleowner-specific variables. So we have to assign them

`so.assign(so_dat)` won't work since `so_dat` isn't grouped by PySAM groups and isn't in the right format (it should be a nested dictionary but is flat instead). So we work around that by exporting it via wrap-export.

If you would add more details, I'd be more than happy to add this to the docs.

**Comment by frohro** [comment]
Thanks for the help Darice!
It is nice to know that on the Mac with pycharm it just works.  I'll fiddle with it a bit more.  In any case, this isn't something you should worry about.  :-)
Your previous comments and these helped me to better understand that having the correct model sequence is really important.  Thanks!

**Comment by frohro** [comment]
I'm checking the example files in the docs, and I found a problem with this same one.  When I generate the JSON file with the SAM 2020.2.17 beta, I get this:
runfile('/home/frohro/AWR/Solar/pysam/Examples/GenericSystemSingleOwnerExample.py', wdir='/home/frohro/AWR/Solar/pysam/Examples')
[autoreload of decorator failed: Traceback (most recent call last):
  File "/home/frohro/.local/lib/python3.7/site-packages/IPython/extensions/autoreload.py", line 245, in check
    superreload(m, reload, self.old_objects)
  File "/home/frohro/.local/lib/python3.7/site-packages/IPython/extensions/autoreload.py", line 450, in superreload
    update_generic(old_obj, new_obj)
  File "/home/frohro/.local/lib/python3.7/site-packages/IPython/extensions/autoreload.py", line 387, in update_generic
    update(a, b)
  File "/home/frohro/.local/lib/python3.7/site-packages/IPython/extensions/autoreload.py", line 357, in update_class
    update_instances(old, new)
  File "/home/frohro/.local/lib/python3.7/site-packages/IPython/extensions/autoreload.py", line 315, in update_instances
    if hasattr(obj, 'items') or (hasattr(obj, '__contains__')
  File "/home/frohro/anaconda3/lib/python3.7/ctypes/__init__.py", line 434, in __getattr__
    dll = self._dlltype(name)
  File "/home/frohro/anaconda3/lib/python3.7/ctypes/__init__.py", line 364, in __init__
    self._handle = _dlopen(self._name, mode)
OSError: items: cannot open shared object file: No such file or directory
]
Traceback (most recent call last):

  File "/home/frohro/AWR/Solar/pysam/Examples/GenericSystemSingleOwnerExample.py", line 26, in <module>
    so.execute()

error: singleowner execution error. system_pre_curtailment_kwac not assigned

When I generate the JSON using the 2018.11.11 stable release, I get this:
runfile('/home/frohro/AWR/Solar/pysam/Examples/GenericSystemSingleOwnerExample.py', wdir='/home/frohro/AWR/Solar/pysam/Examples')
Traceback (most recent call last):

  File "/home/frohro/AWR/Solar/pysam/Examples/GenericSystemSingleOwnerExample.py", line 17, in <module>
    so_dat = dict_to_ssc_table(dic, "singleowner")

  File "/home/frohro/anaconda3/lib/python3.7/site-packages/PySAM/PySSC.py", line 319, in dict_to_ssc_table
    return dict_to_ssc_table_dat(py_dict, cmod_name, dat)

  File "/home/frohro/anaconda3/lib/python3.7/site-packages/PySAM/PySSC.py", line 368, in dict_to_ssc_table_dat
    ssc.data_set_array(dat, ssc_input_data_name.encode("ascii"), py_dict[ssc_input_data_name])

  File "/home/frohro/anaconda3/lib/python3.7/site-packages/PySAM/PySSC.py", line 64, in data_set_array
    count = len(parr)

TypeError: object of type 'float' has no len()
When I generate the JSON with the 2020.1.17 beta, and it works!  

It would be nice if the examples would work in all the systems.

**Comment by frohro** [comment]
I think I discovered the reason that it works on some versions and not on others.  I noticed that sscpi.h and ssc.so are saved to the same directory you save the JSON file of inputs.  If your python script is not in the same directory as the version of those files created by the SAM program, you get the behavior noted just above this.  So I need to change the documentation to make it clear that when you import a JSON file, you need to make sure that you connect to the same directory that the JSON file is found with your python script (assuming that will work like I think).  I haven't tested if this is happening because PySAM is finding the wrong version of ssc.so and sscpi.h before it  finds the correct one, or if it would fail if there is no ssc.so and sscpi.h in the directory with the python script.

**Comment by dguittet** [comment]
There were some code generator issues that were fixed in the latest SAM release, 2.29.r2, which also includes an updated code generator for PySAM.
